compiler/rustc_borrowck/src/diagnostics/region_name.rs

```diff
@@ -683,47 +683,49 @@ impl<'tcx> MirBorrowckCtxt<'_, 'tcx> {
                     }
                     hir::FnRetTy::Return(hir_ty) => (fn_decl.output.span(), Some(hir_ty)),
                 };
-                let mir_description = match hir.body(body).coroutine_kind {
-                    Some(hir::CoroutineKind::Desugared(hir::CoroutineDesugaring::Async, src)) => {
-                        match src {
-                            hir::CoroutineSource::Block => " of async block",
-                            hir::CoroutineSource::Closure => " of async closure",
-                            hir::CoroutineSource::Fn => {
-                                let parent_item =
-                                    tcx.hir_node_by_def_id(hir.get_parent_item(mir_hir_id).def_id);
-                                let output = &parent_item
-                                    .fn_decl()
-                                    .expect("coroutine lowered from async fn should be in fn")
-                                    .output;
-                                span = output.span();
-                                if let hir::FnRetTy::Return(ret) = output {
-                                    hir_ty = Some(self.get_future_inner_return_ty(*ret));
-                                }
-                                " of async function"
+                let mir_description = match kind {
+                    hir::ClosureKind::Coroutine(
+                        hir::CoroutineKind::Desugared(hir::CoroutineDesugaring::Async, src),
+                        _,
+                    ) => match src {
```

Comment:
Can this be flattened to match on the desugaring and the source at the same time?